src/extended-iterable.ts

```diff
@@ -0,0 +1,915 @@
+
+		const handleError = (err: unknown) => {
+			iterator.throw?.(err);
+			throw err;
+		};
+
+		let result: IteratorResult<T> | Promise<IteratorResult<T>>;
+
+		try {
+			result = iterator.next();
+		} catch (err) {
+			return handleError(err);
+		}
+
+		// if first result is async, switch to async mode immediately
+		if (result instanceof Promise) {
+			return this.#asyncAsArray(array, result);
+		}
+
+		// synchronous processing
+		while (!result.done) {
+			array.push(result.value);
+
+			try {
+				result = iterator.next();
+			} catch (err) {
+				return handleError(err);
+			}
+
+			// if iterator becomes async mid-iteration, switch to async mode
+			if (result instanceof Promise) {
+				return this.#asyncAsArray(array, result);
+			}
+		}
+
+		return array;
+	}
+
+	/**
+	 * Helper function to return the async iterator results as an array.
+	 *
+	 * @param array - The array to collect the results in.
+	 * @param currentResult - The current result of the iterator.
+	 * @returns The iterator results as an array.
+	 */
+	async #asyncAsArray(array: T[], currentResult: IteratorResult<T> | Promise<IteratorResult<T>>): Promise<Array<T>> {
+		const iterator = this.#iterator;
+
+		try {
+			let result = await currentResult;
+
+			while (!result.done) {
+				array.push(result.value);
+				result = await iterator.next();
+			}
+
+			return array;
+		} catch (err) {
+			if (iterator.throw) {
+                const throwResult = iterator.throw(err);
+                if (throwResult instanceof Promise) {
+                    return throwResult.then(() => { throw err; });
+                }
+            }
+            throw err;
+		}
+	}
+
+	/**
+	 * Returns the item at the given index.
+	 *
+	 * @param index - The index of the item to return.
+	 * @returns The item at the given index.
+	 *
+	 * @example
+	 * ```typescript
+	 * const iterator = new ExtendedIterable([1, 2, 3]);
+	 * const item = iterator.at(0);
+	 * ```
+	 */
+	at(index: number): T | Promise<T | undefined> | undefined {
+		const iterator = this.#iterator;
+		const handleError = (err: unknown) => {
+			const throwResult = iterator.throw?.(err);
+			if (throwResult instanceof Promise) {
+				return throwResult.then(() => { throw err; });
+			}
+			throw err;
+		};
+
+		let currentIndex = 0;
+		let result: IteratorResult<T> | Promise<IteratorResult<T>>;
+
+		try {
+			if (typeof index !== 'number') {
+				throw new TypeError('index is not a number');
+			}
+			if (index < 0) {
+				throw new RangeError('index must be a positive number');
+			}
+
+			result = iterator.next();
+		} catch (err) {
+			return handleError(err);
+		}
+
+		// if first result is async, handle everything async
+		if (result instanceof Promise) {
+			return this.#asyncAt(result, index, currentIndex).catch(handleError);
+		}
+
+		// synchronous path
+		while (!result.done) {
+			if (currentIndex === index) {
+				if (iterator.return) {
+					return (iterator.return(result.value) as IteratorReturnResult<T>).value;
```

Comment:
Seeing a type cast here seems like a code smell to me. I'll try pulling this locally another time and taking a closer look to understand it better. You don't have to change anything here necessarily, I trust you have a good reason for it.

Reply:
Another TypeScript issues. `iterator.return()` can return `T` or `Promise<T>` depending if we're chained to a sync `Iterator` or `AsyncIterator`. It's impossible for an async iterator to reach this line as it branches above to the `#asyncAt()` flow. The problem is I couldn't find a way to convince TypeScript that `result.value` is indeed `T`. There's no way to detect if `iterator` is sync or async until you call `next()`, `return()`, or `throw()`.

Consumers won't need to worry about this as for-of and for-await-of will get the appropriate iterator.

So it was either cast it as `IteratorReturnResult<T>` or capture the return value from `return()`, then test to see if it's a promise and then await it and return the value. I opted to avoid the bloat.